// common/col.js
class Col{
   /** @type {string} */ table_catalog
   /** @type {string} */ table_schema
   /** @type {string} */ table_name
   /** @type {string} */ column_name
   /** @type {boolean} */ is_primary_key
   /** @type {boolean} */ is_nullable
   /** @type {number} */ ordinal_position
   /** @type {any} */ column_default
   /** @type {'number' | 'varchar' | 'date'} */ data_type

   /** @param {import("common/col").DbCol} col_data */
   constructor(col_data){
      this.table_catalog = col_data.table_catalog
      this.table_schema = col_data.table_schema
      this.table_name = col_data.table_name
      this.column_name = col_data.column_name
      this.is_primary_key = col_data.is_identity === 'YES'
      this.is_nullable = col_data.is_nullable === 'YES'
      this.ordinal_position = col_data.ordinal_position
      this.column_default = col_data.column_default
      this.data_type = defineDataType(col_data.data_type, col_data.udt_name)
   }
}

/**
 * @param {string} data_type 
 * @param {string} udt_name 
 * @returns {'number' | 'varchar' | 'date'}
 */
function defineDataType(data_type, udt_name){
   if (data_type.includes('timestamp')) return 'date'
   if (udt_name.includes('int') || udt_name.includes('float') || udt_name.includes('numeric') ||
         data_type.includes('real') || data_type.includes('numeric')){
      return 'number'
   }
   return 'varchar'
}

module.exports = {Col}
